### PR TITLE
Cpp stereoisomer enumeration

### PR DESCRIPTION
**Pull request by whosayn**
#### Reference Issue
Port EnumerateStereoisomers to C++ (#2110)

#### What does this implement/fix? Explain your changes.
Adds the ability to enumerate stereoisomers to the MolEnumerator library

Todos:
* Update docs about new parameter to enumerate and new enumeration op type
* Port EnumerateStereoisomers.py tests to C++
* Update python wrappers for new api

### CODE REVIEW

**Comment by greglandrum** [comment]
@whosayn : the CI failures here are real

**Comment by whosayn** [comment]
wrapped the new enumerator for python, but still thinking about what to do with the existing python implementation.
Also in the process of porting the old tests for the new C++ enumerator

**Comment by greglandrum** [comment]
@whosayn let me know when you are ready to have this reviewed.

**Comment by whosayn** [comment]
@greglandrum thanks for the reminder

File: Code/GraphMol/MolEnumerator/StereoIsomer.cpp [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    mol.getAtomWithIdx(d_atom)->setChiralTag(next_atom_stereo);
+  }
+  static void add_flippers_from_mol(const ROMol& mol,
+                                    const StereoEnumerationOptions& options,
+                                    std::vector<stereo_flipper_t>& flippers) {
+    if (options.onlyStereoGroups) {
+      return;
+    }
+
+    for (auto* atom : mol.atoms()) {
```
Comment (d-b-w) [inline Code/GraphMol/MolEnumerator/StereoIsomer.cpp:43]: What's the difference between:

    for (auto* atom: mol.atoms()) {
     
and

    for (auto atom: mol.atoms()) {

given that atom is a pointer?

File: Code/GraphMol/MolEnumerator/Wrap/rough_test.py [hunk 2]
```diff
@@ -130,6 +131,30 @@ def testCombined(self):
...
+      self.assertEqual(l, [(1, 'S'), (12, '?'), (14, 'S')])
+
+      enumsi_opt = rdMolEnumerator.StereoEnumerationOptionsV2(maxIsomers=20, onlyUnassigned=False)
+      isomers = list(rdMolEnumerator.EnumerateStereoisomersV2(mol, enumsi_opt))
+      chi_cents = []
+      for iso in isomers:
+        Chem.AssignStereochemistry(iso)
+        chi_cents.append(Chem.FindMolChiralCenters(iso, includeUnassigned=True))
+
+      self.assertEqual(sorted(chi_cents),
```
Comment (d-b-w) [inline Code/GraphMol/MolEnumerator/Wrap/rough_test.py:148]: I know this is annoying, but would you mind making the indentation more coherent here? If clang-format is messing it up, this is a reasonable place to use clang-format: off

File: Code/GraphMol/MolEnumerator/MolEnumerator.h [hunk 3]
```diff
@@ -270,6 +308,23 @@ inline MolBundle enumerate(const ROMol &mol,
   std::vector<MolEnumeratorParams> v = {params};
   return enumerate(mol, v);
 };
+
+struct RDKIT_MOLENUMERATOR_EXPORT StereoEnumerationOptions {
+  bool try_embedding;
```
[the remark this answers is not in the document]
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/MolEnumerator.h]: this was already fixed in the latest diff. Odd that you didn't see it

File: Code/GraphMol/MolEnumerator/StereoIsomer.cpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+  std::vector<MolEnumeratorParams> paramsList;
+
+  MolEnumerator::MolEnumeratorParams stereoParams;
+  stereoParams.dp_operation = MolEnumerator::StereoIsomerOp::createOp();
+  if (options.max_isomers > 0) {
+    stereoParams.maxToEnumerate = options.max_isomers;
+  }
+  paramsList.push_back(stereoParams);
+
+  // NOTE: we don't need the values, so we don't have to worry about underlying
```
[the remark this answers is not in the document]
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/StereoIsomer.cpp]: i updated it to a string and removed this comment in the latest diff. mostly meant to clarify that we didn't need to worry about the underlying const char* being freed.

File: Code/GraphMol/MolEnumerator/StereoIsomer.cpp (hunk 1 above)
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/StereoIsomer.cpp:43]: in this case there's no difference, it's just for readability. there's only a difference for `const auto` vs `const auto*` https://stackoverflow.com/questions/36563017/difference-between-const-auto-and-const-auto

File: Code/GraphMol/MolEnumerator/Wrap/rough_test.py (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/MolEnumerator/Wrap/rough_test.py:148]: In this particular case: we use `yapf` for python code, not clang-format

File: Code/GraphMol/MolEnumerator/MolEnumerator.cpp [hunk 5]
```diff
@@ -48,7 +48,7 @@ void enumerateVariations(std::vector<std::vector<size_t>> &variations,
 
 }  // namespace
 
-namespace detail {
+inline namespace detail {
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/MolEnumerator.cpp:51]: Why this change?

File: Code/GraphMol/MolEnumerator/MolEnumerator.h [hunk 6]
```diff
@@ -237,6 +240,55 @@ class RDKIT_MOLENUMERATOR_EXPORT RepeatUnitOp : public MolEnumeratorOp {
   void initFromMol();
 };
 
+using stereo_flipper_t = std::shared_ptr<StereoFlipper>;
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/MolEnumerator.h:243]: We don't use snake_case. Please change this something in CamelCase. Maybe `StereoFlipperSP`?

File: Code/GraphMol/MolEnumerator/MolEnumerator.h [hunk 7]
```diff
@@ -237,6 +240,55 @@ class RDKIT_MOLENUMERATOR_EXPORT RepeatUnitOp : public MolEnumeratorOp {
...
+
+struct RDKIT_MOLENUMERATOR_EXPORT StereoEnumerationOptions {
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/MolEnumerator.h:245]: Please document the options here.
You can copy the docs from the python file

File: Code/GraphMol/MolEnumerator/MolEnumerator.h [hunk 8]
```diff
@@ -237,6 +240,55 @@ class RDKIT_MOLENUMERATOR_EXPORT RepeatUnitOp : public MolEnumeratorOp {
...
+  bool tryEmbedding = false;
+  bool onlyUnassigned = true;
+  bool onlyStereoGroups = false;
+  bool unique = true;
+  unsigned int maxIsomers = 1024;
+};
+
+RDKIT_MOLENUMERATOR_EXPORT unsigned int get_stereoisomer_count(
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/MolEnumerator.h:253]:
```suggestion
RDKIT_MOLENUMERATOR_EXPORT unsigned int getStereoisomerCount(
```
Same suggestion for ay other snakeCase

File: Code/GraphMol/MolEnumerator/Wrap/rdMolEnumerator.cpp [hunk 9]
```diff
@@ -98,4 +117,54 @@ BOOST_PYTHON_MODULE(rdMolEnumerator) {
 Limitations:
   - Overlapping SRUs, i.e. where one monomer is contained within another, are
     not supported)DOC");
+
+  python::class_<MolEnumerator::StereoEnumerationOptions>(
+      "StereoEnumerationOptionsV2", "Stereoisomer enumeration parameters",
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/Wrap/rdMolEnumerator.cpp:122]: Why V2?

File: Code/GraphMol/MolEnumerator/Wrap/rough_test.py [hunk 10]
```diff
@@ -11,6 +11,7 @@
 import unittest
 
 from rdkit import Chem, RDConfig
+from rdkit.Chem import AllChem
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/Wrap/rough_test.py:14]: I don't think you need this

File: Code/GraphMol/MolEnumerator/MolEnumerator.h [hunk 11]
```diff
@@ -263,12 +315,15 @@ enumerate(const ROMol &mol, const std::vector<MolEnumeratorParams> &paramsList);
 \param maxPerOperation: the maximum number of molecules which an individual
 operation is allowed to generate
 
+\param enumerateStereo: whether to enumerate stereoisomers
```
Comment (greglandrum) [inline Code/GraphMol/MolEnumerator/MolEnumerator.h:318]: Since you added the stereo enumeration as another kind of `MolEnumeratorOp`, why is this parameter needed?

**Review by greglandrum — CHANGES_REQUESTED** [review]
I didn't get into the implementation yet, but here are some API and style suggestions and questoins.

File: Code/GraphMol/MolEnumerator/MolEnumerator.cpp (hunk 5 above)
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/MolEnumerator.cpp]: mostly out of laziness. fixed!

File: Code/GraphMol/MolEnumerator/MolEnumerator.h (hunk 6 above)
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/MolEnumerator.h]: fixed!

File: Code/GraphMol/MolEnumerator/MolEnumerator.h (hunk 7 above)
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/MolEnumerator.h:274]: fixed! I mostly copied the python docs as is, but it should be noted the current functionality is incomplete!

File: Code/GraphMol/MolEnumerator/Wrap/rdMolEnumerator.cpp (hunk 9 above)
Comment (whosayn) [reply Code/GraphMol/MolEnumerator/Wrap/rdMolEnumerator.cpp:122]: Wanted to allow for testing without deleting the previous implementation. I'd like to figure out how to handle the random isomer functionality as part of this review
